evm/src/fuzz/invariant/executor.rs

```diff
@@ -0,0 +1,275 @@
+    sync::Arc,
+};
+
+use ethers::{
+    abi::{Abi, Address, Function},
+    prelude::{ArtifactId, U256},
+};
+use proptest::{
+    strategy::{SBoxedStrategy, Strategy, ValueTree},
+    test_runner::{TestCaseError, TestRunner},
+};
+use revm::DatabaseCommit;
+
+use crate::{
+    executor::{inspector::Fuzzer, Executor, RawCallResult},
+    fuzz::{
+        strategies::{
+            build_initial_state, collect_created_contracts, collect_state_from_call,
+            invariant_strat, override_call_strat, EvmFuzzState,
+        },
+        FuzzCase, FuzzedCases,
+    },
+};
+
+use super::{
+    assert_invariants, BasicTxDetails, FuzzRunIdentifiedContracts, InvariantExecutor,
+    InvariantFuzzError, InvariantFuzzTestResult, InvariantTestOptions, RandomCallGenerator,
+};
+
+impl<'a> InvariantExecutor<'a> {
+    /// Instantiates a fuzzed executor EVM given a testrunner
+    pub fn new(
+        executor: &'a mut Executor,
+        runner: TestRunner,
+        setup_contracts: &'a BTreeMap<Address, (String, Abi)>,
+        project_contracts: &'a BTreeMap<ArtifactId, (Abi, Vec<u8>)>,
+    ) -> Self {
+        Self { executor, runner, setup_contracts, project_contracts }
+    }
+
+    /// Fuzzes any deployed contract and checks any broken invariant at `invariant_address`
+    /// Returns a list of all the consumed gas and calldata of every invariant fuzz case
+    pub fn invariant_fuzz(
+        &mut self,
+        invariants: Vec<&Function>,
+        invariant_address: Address,
+        test_contract_abi: &Abi,
+        test_options: InvariantTestOptions,
+    ) -> eyre::Result<Option<InvariantFuzzTestResult>> {
+        let (fuzz_state, targeted_contracts, strat) =
+            self.prepare_fuzzing(invariant_address, test_contract_abi, test_options)?;
+
+        // Stores the consumed gas and calldata of every successful fuzz call.
+        let fuzz_cases: RefCell<Vec<FuzzedCases>> = RefCell::new(Default::default());
+
+        // Stores the latest reason coming from a test call.  If the runner fails, it will hold the
+        // return reason of the failed test.
+        let revert_reason = RefCell::new(None);
+        let reverts = Cell::new(0);
+        let broken_invariants_count = Cell::new(0);
+        let failed_invariants =
+            RefCell::new(invariants.iter().map(|f| (f.name.to_string(), None)).collect());
+
+        let clean_db = self.executor.backend().db.clone();
+        let executor = RefCell::new(&mut self.executor);
+
+        // The strategy only comes with the first `input`. We fill the rest of the `inputs` until
+        // the desired `depth` so we can use the evolving fuzz dictionary during the
+        // run. We need another proptest runner to query for random values.
+        let branch_runner = RefCell::new(self.runner.clone());
+        let _test_error = self
+            .runner
+            .run(&strat, |mut inputs| {
+                // Scenarios where we want to fail as soon as possible.
```

Comment:
I'd like to make this closure a separate function to get rid of the existing levels, but probably not possible because it borrows all the vars declared outside, so nvm